### PR TITLE
Correct `invmod` for unsigned integers and for negative moduli

### LINKED ISSUE #17595: Strange `invmod` results for unsigned types
This behaviour seems undesirable to me

```
julia> invmod(0x2, 0x3)
0xff
```

Maybe arguments to this function should be passed through `signed`.

### PR DESCRIPTION
**Pull request by eschnett**
- add type signature to restrict arguments to Integer
- use type promotion
- correct existing test case for negative modulos
- add test cases
- also update description of `gcdx`, add a few comments to the source code.

Closes #17595.

### CODE REVIEW

File: base/intfuncs.jl [hunk 1]
```diff
@@ -87,13 +92,16 @@ end
...
-    end
-    x < 0 ? abs(m) + x : x
+    (g != 1 || m == 0) && throw(DomainError())
+    # Note that m might be negative here.
+    # For unsigned T, x might be close to typemax; add m to force a wrap-around.
+    r = mod(x + m, m)
+    # The postcondition is: mod(r * n, m) == mod(T(1), m)
+    r
 end
+invmod(n::Integer, m::Integer) = invmod1(promote(n,m)...)
```
Comment (tkelman) [inline base/intfuncs.jl]: invmod1 a typo?

**Comment by eschnett** [comment]
Typo corrected, respective test case added.

File: base/intfuncs.jl [hunk 2]
```diff
@@ -87,13 +92,16 @@ end
...
-function invmod(n, m)
+function invmod{T<:Integer}(n::T, m::T)
     g, x, y = gcdx(n, m)
-    if g != 1 || m == 0
-        error("no inverse exists")
-    end
-    x < 0 ? abs(m) + x : x
+    (g != 1 || m == 0) && throw(DomainError())
+    # Note that m might be negative here.
+    # For unsigned T, x might be close to typemax; add m to force a wrap-around.
```
Comment (simonbyrne) [inline base/intfuncs.jl]: Is this always going to be the case?

Comment (eschnett) [inline base/intfuncs.jl]: Good question. I did not prove this. The previous implementation says `abs(m) + x`, indicating that adding `m` suffices. I also tested for all 8-bit integers where there was no overflow in intermediate results, and adding `m` was sufficient there. The conditions guaranteed by the `gcdx` algorithm indicate that the result `r` will be "near zero" (see the `gcdx` doc string), so I think this is fine.

**Comment by simonbyrne** [comment]
Test failure seems to be due to change of error type:
https://github.com/JuliaLang/julia/blob/da9c28e22f6b9f602110077f7b664febf5b807c9/test/numbers.jl#L2171-L2172

**Comment by simonbyrne** [comment]
LGTM, once the tests are fixed.

**Comment by simonbyrne** [comment]
This test is now the problem:
https://github.com/JuliaLang/julia/blob/36ed21a3015e16ccde2f0210fbbec496fcb98eda/test/examples.jl#L21

**Comment by simonbyrne** [comment]
Thanks.

**Comment by ararslan** [comment]
This seems to have caused #29971.

**Pull request merged**